e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs

```diff
@@ -0,0 +1,502 @@
+
+            ClientProperties deviceTwin = await deviceClient.GetClientPropertiesAsync().ConfigureAwait(false);
+            if (deviceTwin.Writable.TryGetValue(propName, out string propFromCollection))
+            {
+                Assert.AreEqual<string>(propFromCollection, propValue);
+            }
+            else
+            {
+                Assert.Fail("Property not found in ClientProperties");
+            }
+            await deviceClient.CloseAsync().ConfigureAwait(false);
+            await registryManager.CloseAsync().ConfigureAwait(false);
+        }
+
+        private async Task PropertiesWithComponents_DeviceSetsPropertyAndServiceReceivesItAsync(Client.TransportType transport)
+        {
+            var propName = Guid.NewGuid().ToString();
+            var propValue = Guid.NewGuid().ToString();
+
+            TestDevice testDevice = await TestDevice.GetTestDeviceAsync(Logger, _devicePrefix).ConfigureAwait(false);
+            using var registryManager = RegistryManager.CreateFromConnectionString(Configuration.IoTHub.ConnectionString);
+            using var deviceClient = DeviceClient.CreateFromConnectionString(testDevice.ConnectionString, transport);
+
+            var patch = new ClientPropertyCollection
+            {
+                { ComponentName, propName, propValue }
+            };
+            await deviceClient.UpdateClientPropertiesAsync(patch).ConfigureAwait(false);
+            await deviceClient.CloseAsync().ConfigureAwait(false);
+
+            Twin serviceTwin = await registryManager.GetTwinAsync(testDevice.Id).ConfigureAwait(false);
+            Assert.AreEqual<string>(serviceTwin.Properties.Reported[ComponentName][propName].ToString(), propValue);
+
+            Logger.Trace("verified " + serviceTwin.Properties.Reported[ComponentName][propName].ToString() + "=" + propValue);
+        }
+
+        private async Task PropertiesWithComponents_ServiceDoesNotCreateNullPropertyInCollectionAsync(Client.TransportType transport)
+        {
+            var propName1 = Guid.NewGuid().ToString();
+            var propName2 = Guid.NewGuid().ToString();
+            var propEmptyValue = "{}";
+
+            TestDevice testDevice = await TestDevice.GetTestDeviceAsync(Logger, _devicePrefix).ConfigureAwait(false);
+            using var registryManager = RegistryManager.CreateFromConnectionString(Configuration.IoTHub.ConnectionString);
+            using var deviceClient = DeviceClient.CreateFromConnectionString(testDevice.ConnectionString, transport);
+
+            await deviceClient
+                .UpdateClientPropertiesAsync(
+                    new ClientPropertyCollection
+                    {
+                        { ComponentName, propName1, null }
+                    })
+                .ConfigureAwait(false);
+            Twin serviceTwin = await registryManager.GetTwinAsync(testDevice.Id).ConfigureAwait(false);
+            Assert.IsTrue(serviceTwin.Properties.Reported.Contains(ComponentName));
```

Comment:
assertion for component name containing an empty value? we should probably check with service if that is the expected behavior, i.e. to remove an updated property we pass in an empty value, how to remove a component?

Reply:
done